checks.d/kubernetes.py

```diff
@@ -273,11 +332,73 @@ def _update_pods_metrics(self, instance, pods):
             _tags.append('kube_replication_controller:{0}'.format(ctrl))
             self.publish_gauge(self, NAMESPACE + '.pods.running', pod_count, _tags)
+
+    def _process_events(self, instance, pods_list):
+        """
+        Retrieve a list of events from the kubernetes API.
+
+        At the moment (k8s v1.3) there is no support to select events based on a timestamp query, so we
+        go through the whole list every time. This should be fine for now as events
+        have a TTL of one hour[1] but logic needs to improve as soon as they provide
+        query capabilities or at least pagination, see [2][3].
+
+        [1] https://github.com/kubernetes/kubernetes/blob/release-1.3.0/cmd/kube-apiserver/app/options/options.go#L51
+        [2] https://github.com/kubernetes/kubernetes/issues/4432
+        [3] https://github.com/kubernetes/kubernetes/issues/1362
+        """
+        node_ip, node_name = self.kubeutil.get_node_info()
+        self.log.debug('Processing events on {} [{}]'.format(node_name, node_ip))
+
+        k8s_namespace = instance.get('namespace', 'default')
+        events_endpoint = '{}/namespaces/{}/events'.format(self.kubeutil.kubernetes_api_url, k8s_namespace)
+        self.log.debug('Kubernetes API endpoint to query events: %s' % events_endpoint)
+
+        events = self.kubeutil.retrieve_json_auth(events_endpoint, self.kubeutil.get_auth_token())
+        event_items = events.get('items') or []
+        last_read = self.kubeutil.last_event_collection_ts[k8s_namespace]
+        most_recent_read = 0
+
+        self.log.debug('Found {} events, filtering out using timestamp: {}'.format(len(event_items), last_read))
+
+        for event in event_items:
+            involved_obj = event.get('involvedObject', {})
+
+            # skip if the event is too old
+            event_ts = int(time.mktime(time.strptime(event.get('lastTimestamp'), '%Y-%m-%dT%H:%M:%SZ')))
+            if event_ts <= last_read:
+                continue
+
+            # compute the most recently seen event, without relying on items order
+            if event_ts > most_recent_read:
+                most_recent_read = event_ts
+
+            title = '{} {} on {}'.format(involved_obj.get('name'), event.get('reason'), node_name)
+            message = event.get('message')
+            source = event.get('source')
+            if source:
+                message += '\nSource: {} {}\n'.format(source.get('component', ''), source.get('host', ''))
+            msg_body = "%%%\n{}\n```\n{}\n```\n%%%".format(title, message)
+            dd_event = {
+                'timestamp': event_ts,
+                'host': node_ip,
+                'event_type': EVENT_TYPE,
+                'msg_title': title,
+                'msg_text': msg_body,
+                'source_type_name': EVENT_TYPE,
+                'event_object': 'kubernetes:{}'.format(involved_obj.get('name')),
+            }
+            self.log.debug('Posting event: {}'.format(dd_event))
```

Comment:
Yeah it made sense with the first version implementing the event filtering but this way the event collector will be hammered. Better with or without it?